### PR TITLE
[MXNET-265] Update optimizer doc to clarify wd behaviors

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
## Description ##
Updates the doc for the following optimizers:
- FTML
- AdaDelta
- Adam
- Adagrad

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

#9881

### CODE REVIEW

**Comment by eric-haibin-lin** [comment]
@ZiyueHuang @szhengac @sxjscience could you help review?

File: python/mxnet/optimizer.py [hunk 1]
```diff
@@ -1210,6 +1210,15 @@ class AdaDelta(Optimizer):
     This class implements AdaDelta, an optimizer described in  *ADADELTA: An adaptive
     learning rate method*, available at https://arxiv.org/abs/1212.5701.
 
+    This optimizer updates each weight by::
+
+        grad = clip(grad * rescale_grad, clip_gradient)
+        acc_grad = rho * acc_grad + (1. - rho) * grad * grad
+        delta = sqrt(acc_delta + epsilon) / sqrt(acc_grad + epsilon) * grad
+        acc_delta = rho * acc_delta + (1. - rho) * delta * delta
+        weight -= delta + wd * weight
```
Comment (szhengac) [inline python/mxnet/optimizer.py]: See my comment for Adagrad

File: python/mxnet/optimizer.py [hunk 2]
```diff
@@ -1076,6 +1076,13 @@ class AdaGrad(Optimizer):
     Methods for Online Learning and Stochastic Optimization*, and available at
     http://www.jmlr.org/papers/volume12/duchi11a/duchi11a.pdf.
 
+    This optimizer updates each weight by::
+    
+            grad = clip(grad * rescale_grad, clip_gradient)
+            history += square(grad)
+            div = grad / sqrt(history + float_stable_eps)
+            weight += (div + weight * wd) * -lr
```
Comment (szhengac) [inline python/mxnet/optimizer.py]: This step is problematic. For Adagrad, the stpesize is generally much smaller than the one for SGD, but here you separate the gradient of l2 norm from the div. This can make the training slow if we use a small stepwise, vice versa, a large stepwise can lead to divergence

**Comment by szhengac** [comment]
I found the following line rescales the gradient outside the clipping function. This is inconsistent with other optimizers. 

https://github.com/eric-haibin-lin/mxnet/blob/90bb058bccb35df63366057d0538fb411f2f5151/python/mxnet/optimizer.py#L440

File: python/mxnet/optimizer.py [hunk 3]
```diff
@@ -538,7 +538,7 @@ def update_multi_precision(self, index, weight, grad, state):
 class Signum(Optimizer):
     """The Signum optimizer that takes the sign of gradient or momentum.
 
-    The optimizer updates the weight by:
+    The optimizer updates the weight by::
```
Comment (haojin2) [inline python/mxnet/optimizer.py]: an extra ":" ?

**Comment by eric-haibin-lin** [comment]
Merged wd term before clipping grad for AdaGrad.

**Comment by szhengac** [comment]
It also needs to be modified for adadelta 

https://github.com/eric-haibin-lin/mxnet/blob/e05933d9e0f9ad60f99e2c42ce9cfb716766ce79/python/mxnet/optimizer.py#L1261

**Comment by eric-haibin-lin** [comment]
Ah, good point. I'll change that, too.

File: python/mxnet/optimizer.py (hunk 3 above)
Comment (ThomasDelteil) [inline python/mxnet/optimizer.py]: extra `:`

**Comment by eric-haibin-lin** [comment]
I'm still hesitated to change all these behaviors since they're will break existing training scripts..
